### PR TITLE
feat: Save last used stack and order existing deployments by MRU stack

### PR DESCRIPTION
**Pull request by philasmar**
*Issue #, if available:*

*Description of changes:*
feat: Save last used stack and order existing deployments by MRU stack

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/AWS.Deploy.CLI/Commands/DeleteDeploymentCommand.cs [hunk 1]
```diff
@@ -66,6 +69,7 @@ public async Task ExecuteAsync(string stackName)
                 var _ = monitor.StartAsync();
 
                 await WaitForStackDelete(stackName);
+
```
Comment (normj) [inline src/AWS.Deploy.CLI/Commands/DeleteDeploymentCommand.cs:72]: Should we be removing the stack from `aws-deployments.json` after we delete it?

File: src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestModel.cs [hunk 2]
```diff
@@ -11,11 +11,14 @@ namespace AWS.Deploy.Common.DeploymentManifest
     /// </summary>
     public class DeploymentManifestModel
     {
+        public List<LastDeployedStack> LastDeployedStacks { get; set; }
+
         [JsonProperty("deployment-projects")]
         public List<DeploymentManifestEntry> DeploymentManifestEntries { get; set; }
```
Comment (normj) [inline src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestModel.cs:17]: I'm seeing when I use this that `deployment-projects` is getting written to the JSON file as an empty string. I'm assuming `LastDeployedStacks` would also be serialized as an empty array as well if there were no values. Malhar did some work in his previous PR to make sure we weren't writing out empty lists when serializing the recipe. Check with him to see what he did.

File: src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestModel.cs [hunk 3]
```diff
@@ -11,11 +11,14 @@ namespace AWS.Deploy.Common.DeploymentManifest
     /// </summary>
     public class DeploymentManifestModel
     {
+        public List<LastDeployedStack> LastDeployedStacks { get; set; }
```
Comment (normj) [inline src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestModel.cs:14]: JSON property names are inconsistent. This one is `LastDeployedStacks` and deployment projects is `deployment-projects`. When should use either pascal casing or hyphen across the file.

File: src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                    $"for the deployment project stored at '{saveCdkDirectoryFullPath}'", ex);
+            }
+            
+        }
+
+        /// <summary>
+        /// This method updates the deployment manifest json file by adding the name of the stack that was most recently used.
+        /// If the manifest file does not exists then a new file is generated.
+        /// </summary>
+        public async Task UpdateLastDeployedStack(string stackName)
```
Comment (normj) [inline src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs:87]: Should this method also clean up any orphan stacks in the LastDeployedStacks for the current account and region? Or is there some other mechanism to clean up the file for stacks deleted in the console or some other tool.

File: src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        /// </summary>
+        public async Task DeleteLastDeployedStack(string stackName)
+        {
+            try
+            {
+                var deploymentManifestModel = await GetDeploymentManifest();
+                var lastDeployedStack = deploymentManifestModel?.LastDeployedStacks
+                    .FirstOrDefault(x => x.AWSAccountId.Equals(_session.AWSAccountId) && x.AWSRegion.Equals(_session.AWSRegion));
+
+                if (deploymentManifestModel != null && lastDeployedStack != null)
```
Comment (normj) [inline src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs:144]: nit: lastDeployedStack has to be null if deploymentManifestModel so you could just check `lastDeployedStack != null`

File: src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        /// This method parses the deployment-manifest file into a <see cref="DeploymentManifestModel"/>
+        /// </summary>
+        public async Task<DeploymentManifestModel?> GetDeploymentManifest()
+        {
+            var deploymentManifestFilePath = GetDeploymentManifestFilePath(_targetApplicationFullPath);
+
+            if (!_fileManager.Exists(deploymentManifestFilePath))
+                return null;
+            var manifestFilejsonString = await _fileManager.ReadAllTextAsync(deploymentManifestFilePath);
+            return JsonConvert.DeserializeObject<DeploymentManifestModel>(manifestFilejsonString);
```
Comment (normj) [inline src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs:197]: I think if we fail to deserialize `aws-deployments.json` we should log it and then return an empty `DeploymentManifestModel`. We shouldn't fail deployment because somebody dumb fingered this file of mostly optional information. Especially since if we fail to deserialize the error will manifest in probably confusing ways in the toolkit.

File: src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+        }
+
+        /// <summary>
+        /// This method parses the <see cref="DeploymentManifestModel"/> into a string and writes it to disk.
+        /// </summary>
+        private async Task WriteDeploymentManifestFile(DeploymentManifestModel deploymentManifestModel)
+        {
+            var deploymentManifestFilePath = GetDeploymentManifestFilePath(_targetApplicationFullPath);
+            var manifestFileJsonString = JsonConvert.SerializeObject(deploymentManifestModel, Formatting.Indented);
```
Comment (normj) [inline src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs:206]: Use something like [SerializeRecipeContractResolver](https://github.com/aws/aws-dotnet-deploy/blob/dev/src/AWS.Deploy.Common/SerializeRecipeContractResolver.cs) to not serialize empty lists.

File: src/AWS.Deploy.Orchestration/Utilities/DeployedApplicationQueryer.cs [hunk 8]
```diff
@@ -76,6 +89,19 @@ public async Task<List<CloudApplication>> GetExistingDeployedApplications(IList<
...
             }
 
+            if (_deploymentManifestEngine != null &&
+                _session != null)
+            {
+                var deploymentManifest = await _deploymentManifestEngine.GetDeploymentManifest();
+                var lastDeployedStack = deploymentManifest?.LastDeployedStacks
+                    .FirstOrDefault(x => x.AWSAccountId.Equals(_session.AWSAccountId) && x.AWSRegion.Equals(_session.AWSRegion));
+
+                return apps.OrderBy(x => {
```
Comment (normj) [inline src/AWS.Deploy.Orchestration/Utilities/DeployedApplicationQueryer.cs:99]: I believe the plan was to sort the list of the stacks whether they are in or not of the `LastDeployedStacks` and then the stacks last update date. That way the stack you most recently deployed to when show up first. We should also return the last updated date in case the toolkit wants to show it so the users can see the order is date based and not alpha.

**Review by normj — CHANGES_REQUESTED** [review]

File: src/AWS.Deploy.CLI/Commands/DeleteDeploymentCommand.cs (hunk 1 above)
Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/DeleteDeploymentCommand.cs:72]: Originally I was doing that but since modifying the aws-deployments.json file needs knowledge of the project path and the delete command does not have it, I decided against it. The alternative would be to modify the delete command to only run in a directory with a valid project (which i don't think we want to do)

Comment (normj) [reply src/AWS.Deploy.CLI/Commands/DeleteDeploymentCommand.cs:72]: Seems like we could do a best attempt. If the current directory or if `--project-path` is set then we can check the local deployment file. If the stack is not listed or the file is not found then we don't do any clean up. Currently the delete command does still take in `--project-path`.

Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/DeleteDeploymentCommand.cs:75]: Done

File: src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestModel.cs (hunk 2 above)
Comment (philasmar) [reply src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestModel.cs]: Fixed

File: src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs (hunk 5 above)
Comment (philasmar) [reply src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs]: If i dont explicitly check deploymentManifestModel for null, I will get a Possible null reference error.

File: src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs (hunk 6 above)
Comment (philasmar) [reply src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs]: Fixed

File: src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs (hunk 7 above)
Comment (philasmar) [reply src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs]: Fixed

File: src/AWS.Deploy.Orchestration/Utilities/DeployedApplicationQueryer.cs (hunk 8 above)
Comment (philasmar) [reply src/AWS.Deploy.Orchestration/Utilities/DeployedApplicationQueryer.cs]: Fixed

File: src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+            try
+            {
+                if (!_directoryManager.Exists(saveCdkDirectoryFullPath))
+                    return;
+
+                var deploymentManifestModel = await GetDeploymentManifest();
+                var targetApplicationDirectoryPath = _directoryManager.GetDirectoryInfo(_targetApplicationFullPath).Parent.FullName;
+                var saveCdkDirectoryRelativePath = _directoryManager.GetRelativePath(targetApplicationDirectoryPath, saveCdkDirectoryFullPath);
+
+                if (deploymentManifestModel != null && deploymentManifestModel.DeploymentManifestEntries != null)
```
Comment (normj) [inline src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs:65]: If `deploymentManifestModel` is not null then you should always use that object not just if `deploymentManifestModel.DeploymentManifestEntries` is not null. Otherwise you are wiping out the rest of the data in the `deploymentManifestModel` the first time something is added to `DeploymentManifestEntries`.

File: src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+                if (_session.AWSAccountId == null || _session.AWSRegion == null)
+                    throw new FailedToUpdateDeploymentManifestFileException("The AWS Account Id or Region is not defined.");
+
+                var deploymentManifestModel = await GetDeploymentManifest();
+                var lastDeployedStack = deploymentManifestModel?.LastDeployedStacks?
+                    .FirstOrDefault(x => x.AWSAccountId.Equals(_session.AWSAccountId) && x.AWSRegion.Equals(_session.AWSRegion));
+
+                if (deploymentManifestModel != null && lastDeployedStack != null)
+                {
+                    if (lastDeployedStack != null)
```
Comment (normj) [inline src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs:102]: Since we are sorting by the CloudFormation update timestamp we don't need to keep changing the order of this file. I think that will cause frustration with a pair of developers both deploying to their own stacks in the same account but the file keeps getting changes and then being listed as a changed file for change control. If we wanted to have a deterministic  list just always keep the list in alpha order.

File: src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+        /// This method parses the <see cref="DeploymentManifestModel"/> into a string and writes it to disk.
+        /// </summary>
+        private async Task WriteDeploymentManifestFile(DeploymentManifestModel deploymentManifestModel)
+        {
+            var deploymentManifestFilePath = GetDeploymentManifestFilePath(_targetApplicationFullPath);
+            var manifestFileJsonString = JsonConvert.SerializeObject(deploymentManifestModel, new JsonSerializerSettings
+            {
+                Formatting = Formatting.Indented,
+                NullValueHandling = NullValueHandling.Ignore,
+                ContractResolver = new SerializeRecipeContractResolver()
```
Comment (normj) [inline src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs:247]: Should we change the name of this resolver class to something more generic since this is not a recipe file.

File: src/AWS.Deploy.CLI/Commands/DeployCommand.cs [hunk 12]
```diff
@@ -18,6 +18,7 @@
 using AWS.Deploy.Orchestration.Data;
```
[the remark this answers is not in the document]
Comment (normj) [reply src/AWS.Deploy.CLI/Commands/DeployCommand.cs:18]: Looking at this a bit more it seems like we have a divergence between how the CLI is determining the list and ServerMode. ServerMode is calling `GetExistingDeployedApplications` with the list of possible recommendations to filter. The CLI is calling `GetExistingDeployedApplications` without passing in a list of recommendations to filter and then calling a separate `GetCompatibleApplications` method to filter which is creating a new list and disrupting your sorting. We need to resolve this logic to have a single solution otherwise we are going to have bugs here.

**Review by normj — CHANGES_REQUESTED** [review]

File: src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestModel.cs [hunk 13]
```diff
@@ -11,11 +11,13 @@ namespace AWS.Deploy.Common.DeploymentManifest
     /// </summary>
     public class DeploymentManifestModel
     {
-        [JsonProperty("deployment-projects")]
-        public List<DeploymentManifestEntry> DeploymentManifestEntries { get; set; }
+        public List<LastDeployedStack>? LastDeployedStacks { get; set; }
 
-        public DeploymentManifestModel(List<DeploymentManifestEntry> deploymentManifestEntries)
+        public List<DeploymentManifestEntry>? DeploymentManifestEntries { get; set; }
```
[the remark this answers is not in the document]
Comment (96malhar) [reply src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestModel.cs:16]: We already have JsonProperty called "deployment-projects". But its being removed as part of this PR

[the remark this answers is not in the document]
Comment (normj) [reply src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestModel.cs:16]: @96malhar Yes, I asked in previous comment to make the name follow the same casing pattern. I'm good with using deployment project but we should probably use `DeploymentProjects` instead of `deployment-projects`.

[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestModel.cs]: Fixed

File: src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestModel.cs [hunk 14]
```diff
@@ -11,11 +11,13 @@ namespace AWS.Deploy.Common.DeploymentManifest
     /// </summary>
     public class DeploymentManifestModel
     {
-        [JsonProperty("deployment-projects")]
-        public List<DeploymentManifestEntry> DeploymentManifestEntries { get; set; }
+        public List<LastDeployedStack>? LastDeployedStacks { get; set; }
```
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestModel.cs:13]: Fixed

File: src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs (hunk 11 above)
Comment (philasmar) [reply src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs]: Done

File: src/AWS.Deploy.CLI/Commands/CommandFactory.cs [hunk 15]
```diff
@@ -171,6 +168,14 @@ private Command BuildDeployCommand()
 
                     var dockerEngine = new DockerEngine.DockerEngine(projectDefinition);
 
+                    var targetApplicationFullPath = _directoryManager.GetDirectoryInfo(projectDefinition.ProjectPath).FullName;
+
+                    var deploymentManifestEngine = new DeploymentManifestEngine(_directoryManager, _fileManager, session, targetApplicationFullPath);
+
+                    var deployedApplicationQueryer = new DeployedApplicationQueryer(_awsResourceQueryer, deploymentManifestEngine, session, _orchestratorInteractiveService);
+
+                    var customRecipeLocator = new CustomRecipeLocator(deploymentManifestEngine, _orchestratorInteractiveService, _commandLineWrapper, _directoryManager);
```
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/CommandFactory.cs]: Fixed

File: src/AWS.Deploy.CLI/Commands/CommandFactory.cs [hunk 16]
```diff
@@ -256,7 +262,27 @@ private Command BuildDeleteCommand()
...
+                    {
+                        var projectDefinition = await _projectParserUtility.Parse(input.ProjectPath ?? "");
+
+                        var callerIdentity = await _awsResourceQueryer.GetCallerIdentity();
+
+                        var session = new OrchestratorSession(
+                        projectDefinition,
+                        awsCredentials,
+                        awsRegion,
+                        callerIdentity.Account);
```
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/CommandFactory.cs]: Fixed

File: src/AWS.Deploy.CLI/Commands/CommandFactory.cs [hunk 17]
```diff
@@ -256,7 +262,27 @@ private Command BuildDeleteCommand()
...
+
+                        var targetApplicationFullPath = _directoryManager.GetDirectoryInfo(projectDefinition.ProjectPath).FullName;
+
+                        deploymentManifestEngine = new DeploymentManifestEngine(_directoryManager, _fileManager, session, targetApplicationFullPath);
+                    }
+                    catch (FailedToFindDeployableTargetException) { }
```
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/CommandFactory.cs:281]: This is to address Norm's comment. We want to do a best effort in deleting the stack from the file, but if the user does not provide the project path as part of the delete command, then we cannot do this. This is why we catch the exception.

File: test/AWS.Deploy.CLI.IntegrationTests/SaveCdkDeploymentProject/SaveCdkDeploymentProjectTests.cs [hunk 18]
```diff
@@ -14,52 +15,60 @@ namespace AWS.Deploy.CLI.IntegrationTests.SaveCdkDeploymentProject
     [Collection("SaveCdkDeploymentProjectTests")]
     public class SaveCdkDeploymentProjectTests : IDisposable
     {
-        private readonly string _targetApplicationProjectPath;
-        private readonly string _deploymentManifestFilePath;
+        //private readonly string _targetApplicationProjectPath;
+        //private readonly string _deploymentManifestFilePath;
```
[the remark this answers is not in the document]
Comment (philasmar) [reply test/AWS.Deploy.CLI.IntegrationTests/SaveCdkDeploymentProject/SaveCdkDeploymentProjectTests.cs]: Done

File: test/AWS.Deploy.CLI.IntegrationTests/SaveCdkDeploymentProject/SaveCdkDeploymentProjectTests.cs [hunk 19]
```diff
@@ -14,52 +15,60 @@ namespace AWS.Deploy.CLI.IntegrationTests.SaveCdkDeploymentProject
...
         private string _saveDirectoryPath;
+        private string _stackName;
 
         public SaveCdkDeploymentProjectTests()
         {
             var testAppsDirectoryPath = Utilities.ResolvePathToTestApps();
-            _targetApplicationProjectPath = Path.Combine(testAppsDirectoryPath, "WebAppWithDockerFile");
-            _deploymentManifestFilePath = Path.Combine(testAppsDirectoryPath, "WebAppWithDockerFile", "aws-deployments.json");
+            //_targetApplicationProjectPath = Path.Combine(testAppsDirectoryPath, "WebAppWithDockerFile");
+            //_deploymentManifestFilePath = Path.Combine(testAppsDirectoryPath, "WebAppWithDockerFile", "aws-deployments.json");
```
[the remark this answers is not in the document]
Comment (philasmar) [reply test/AWS.Deploy.CLI.IntegrationTests/SaveCdkDeploymentProject/SaveCdkDeploymentProjectTests.cs]: Done

File: src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+
+                var deploymentManifestModel = await GetDeploymentManifest();
+                var lastDeployedStack = deploymentManifestModel?.LastDeployedStacks?
+                    .FirstOrDefault(x => x.AWSAccountId.Equals(_session.AWSAccountId) && x.AWSRegion.Equals(_session.AWSRegion));
+
+                if (deploymentManifestModel != null)
+                {
+                    if (lastDeployedStack != null)
+                    {
+                        if (lastDeployedStack.Stacks == null)
```
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs:107]: Since we are ignoring empty lists when serializing, this can be null when serializing.

**Comment by ganeshnj** [comment]
General question: what is the rational behind storing the timestamp in aws-deployments.json file than using `LastUpdatedTime` to order all the stacks deployed through deploy tool?

This approach will be more accurate in case stack gets updated out of the lifecycle of deploy tool and requires less maintenance on client side.

**Review by ganeshnj — CHANGES_REQUESTED** [review]

**Comment by ganeshnj** [comment]
Another downside of this approach is, given aws-deployments.json file is source controlled, as soon as multiple developers will commit their changes, this logic will not work.

By keeping stack tracking per system (somewhere in temp location) would solve this problem.

File: src/AWS.Deploy.Orchestration/Utilities/DeployedApplicationQueryer.cs [hunk 21]
```diff
@@ -73,7 +87,32 @@ public async Task<List<CloudApplication>> GetExistingDeployedApplications(IList<
                     continue;
                 }
 
-                apps.Add(new CloudApplication(stack.StackName, recipeId));
+                apps.Add(new CloudApplication(stack.StackName, recipeId, stack.LastUpdatedTime));
```
Comment (96malhar) [inline src/AWS.Deploy.Orchestration/Utilities/DeployedApplicationQueryer.cs:90]: What time zone is `stack.LastUpdatedTime` returned in? Also, as per the documentation, this field will only be returned if the stack has been updated at least once. 

image

source - https://docs.aws.amazon.com/sdkfornet/v3/apidocs/items/CloudFormation/TStack.html

Will this be a problem?

File: src/AWS.Deploy.CLI/Commands/CommandHandlerInput/GenerateDeploymentProjectCommandHandlerInput.cs [hunk 22]
```diff
@@ -8,6 +8,8 @@ namespace AWS.Deploy.CLI.Commands.CommandHandlerInput
     /// </summary>
     public class GenerateDeploymentProjectCommandHandlerInput
     {
+        public string? Profile { get; set; }
+        public string? Region { get; set; }
```
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/CommandHandlerInput/GenerateDeploymentProjectCommandHandlerInput.cs]: Fixed. No need for AWS creds for this command.

File: src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs [hunk 23]
```diff
@@ -0,0 +1,282 @@
...
+        Task DeleteLastDeployedStack(string stackName, string? awsAccountId, string? awsRegion, string targetApplicationFullPath);
+        Task<List<string>> GetRecipeDefinitionPaths(string targetApplicationFullPath);
+        Task CleanOrphanStacks(List<string> deployedStacks, string? awsAccountId, string? awsRegion, string targetApplicationFullPath);
+    }
+
+    /// <summary>
+    /// This class contains the helper methods to update the deployment manifest file
+    /// that keeps track of the save CDK deployment projects.
+    /// </summary>
+    public class DeploymentManifestEngine : IDeploymentManifestEngine
```
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.Orchestration/DeploymentManifestEngine.cs:30]: Added tests

File: src/AWS.Deploy.CLI/Commands/DeployCommand.cs [hunk 24]
```diff
@@ -131,24 +140,23 @@ public async Task<(Orchestrator, Recommendation, CloudApplication)> InitializeDe
...
                     _customRecipeLocator,
-                    new List<string> { RecipeLocator.FindRecipeDefinitionsPath() });
+                    new List<string> { RecipeLocator.FindRecipeDefinitionsPath() },
+                    _directoryManager);
 
             // Determine what recommendations are possible for the project.
             var recommendations = await GenerateDeploymentRecommendations(orchestrator, deploymentProjectPath);
 
-            // Get all existing applications that were previously deployed using our deploy tool.
-            var allDeployedApplications = await _deployedApplicationQueryer.GetExistingDeployedApplications();
```
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/DeployCommand.cs:142]: I added that back to get all stacks.

File: src/AWS.Deploy.CLI/Commands/DeployCommand.cs [hunk 25]
```diff
@@ -131,24 +140,23 @@ public async Task<(Orchestrator, Recommendation, CloudApplication)> InitializeDe
...
-
             // Filter compatible applications that can be re-deployed  using the current set of recommendations.
-            var compatibleApplications = GetCompatibleApplications(allDeployedApplications, recommendations);
+            var compatibleApplications = await _deployedApplicationQueryer.GetExistingDeployedApplications(recommendations, _session);
 
             // Get Cloudformation stack name.
             var cloudApplicationName = GetCloudApplicationName(stackName, userDeploymentSettings, compatibleApplications);
 
             // Find existing application with the same CloudFormation stack name.
-            var deployedApplication = allDeployedApplications.FirstOrDefault(x => string.Equals(x.Name, cloudApplicationName));
```
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/DeployCommand.cs]: Fixed.

File: src/AWS.Deploy.CLI/Commands/DeployCommand.cs [hunk 26]
```diff
@@ -180,23 +188,6 @@ public async Task<(Orchestrator, Recommendation, CloudApplication)> InitializeDe
...
-        {
-            var compatibleApplications = new List<CloudApplication>();
-            foreach (var app in allDeployedApplications)
-            {
-                if (recommendations.Any(rec => string.Equals(rec.Recipe.Id, app.RecipeId, StringComparison.Ordinal)))
-                    compatibleApplications.Add(app);
-            }
-            return compatibleApplications;
-        }
-
```
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/DeployCommand.cs:199]: This is not needed. It is done as part of GetExistingDeployedApplications

File: src/AWS.Deploy.CLI/Extensions/CustomServiceCollectionExtension.cs [hunk 27]
```diff
@@ -5,12 +5,12 @@
 using AWS.Deploy.CLI.Commands.TypeHints;
 using AWS.Deploy.CLI.Utilities;
 using AWS.Deploy.Common;
-using AWS.Deploy.Common.DeploymentManifest;
 using AWS.Deploy.Common.Extensions;
 using AWS.Deploy.Common.IO;
 using AWS.Deploy.Orchestration;
 using AWS.Deploy.Orchestration.CDK;
 using AWS.Deploy.Orchestration.Data;
+using AWS.Deploy.Orchestration.DeploymentManifest;
```
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.CLI/Extensions/CustomServiceCollectionExtension.cs:13]: DeploymentManifestEntry and DeploymentManifestModel are POCO classes and could be used by other projects. They dont NEED to be in the orchestrator.

File: src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs [hunk 28]
```diff
@@ -242,13 +244,15 @@ public async Task<IActionResult> GetExistingDeployments(string sessionId)
             var output = new GetExistingDeploymentsOutput();
 
             var deployedApplicationQueryer = serviceProvider.GetRequiredService<IDeployedApplicationQueryer>();
-            state.ExistingDeployments = await deployedApplicationQueryer.GetExistingDeployedApplications(state.NewRecommendations);
+            var session = CreateOrchestratorSession(state);
+            state.ExistingDeployments = await deployedApplicationQueryer.GetExistingDeployedApplications(state.NewRecommendations, session);
```
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.CLI/ServerMode/Controllers/DeploymentController.cs:248]: This is what this API is meant to do.

File: src/AWS.Deploy.Orchestration/Utilities/DeployedApplicationQueryer.cs (hunk 21 above)
Comment (philasmar) [reply src/AWS.Deploy.Orchestration/Utilities/DeployedApplicationQueryer.cs:90]: I checked and every stack deployed by the deploy tool has the Last Updated Time. I also tried to fetch them with the API and it was fine.

File: src/AWS.Deploy.CLI/Commands/DeployCommand.cs (hunk 26 above)
[the remark this answers is not in the document]
Comment (96malhar) [reply src/AWS.Deploy.CLI/Commands/DeployCommand.cs:199]: Move this method to `DeployedApplicationQueryer`  and use this to filter the `compatibleApplications` from `allDeployedApplications`. This way we dont need to make a second API call to CloudFormation.

File: test/AWS.Deploy.CLI.IntegrationTests/SaveCdkDeploymentProject/CustomRecipeLocatorTests.cs [hunk 29]
```diff
@@ -51,7 +52,7 @@ public CustomRecipeLocatorTests()
             _solutionDirectoryPath = directoryManager.GetDirectoryInfo(solutionPath).Parent.FullName;
         }
 
-        [Fact]
+        [Fact(Skip = "Needs Fixing")]
```
[the remark this answers is not in the document]
Comment (philasmar) [reply test/AWS.Deploy.CLI.IntegrationTests/SaveCdkDeploymentProject/CustomRecipeLocatorTests.cs]: Done

File: src/AWS.Deploy.CLI/Commands/DeployCommand.cs (hunk 26 above)
[the remark this answers is not in the document]
Comment (philasmar) [reply src/AWS.Deploy.CLI/Commands/DeployCommand.cs:199]: Done

File: src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestEngine.cs [hunk 30]
```diff
@@ -57,7 +57,14 @@ public async Task UpdateDeploymentManifestFile(string saveCdkDirectoryFullPath,
                 if (_fileManager.Exists(deploymentManifestFilePath))
                 {
                     deploymentManifestModel = await ReadManifestFile(deploymentManifestFilePath);
-                    deploymentManifestModel.DeploymentManifestEntries.Add(new DeploymentManifestEntry(saveCdkDirectoryRelativePath));
+                    if (deploymentManifestModel.DeploymentProjects == null)
+                    {
+                        deploymentManifestModel.DeploymentProjects = new List<DeploymentManifestEntry> { new DeploymentManifestEntry(saveCdkDirectoryRelativePath) };
+                    }
```
Comment (ganeshnj) [inline src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestEngine.cs:63]: This and L100 check look redundant, `DeploymentManifestModel` has constructor which always initializes with a non-null List<DeploymentManifestEntry>.

File: src/AWS.Deploy.Orchestration/LocalUserSettings/LocalUserSettingsEngine.cs [hunk 31]
```diff
@@ -0,0 +1,208 @@
...
+
+                var localUserSettings = await GetLocalUserSettings();
+                var lastDeployedStack = localUserSettings?.LastDeployedStacks?
+                    .FirstOrDefault(x => x.Exists(awsAccountId, awsRegion, projectName));
+
+                if (localUserSettings != null)
+                {
+                    if (lastDeployedStack != null)
+                    {
+                        if (lastDeployedStack.Stacks == null)
```
Comment (ganeshnj) [inline src/AWS.Deploy.Orchestration/LocalUserSettings/LocalUserSettingsEngine.cs:57]: It is redundant check, `LastDeployedStack` has constructor that initializes Stacks with a non-nullable List<string>

File: src/AWS.Deploy.Common/CloudApplication.cs [hunk 32]
```diff
@@ -27,16 +29,27 @@ public class CloudApplication
...
 
+        /// <summary>
+        /// Last updated time of CloudFormation stack
+        /// </summary>
+        public DateTime? LastUpdatedTime { get; set; }
+
+        /// <summary>
+        /// Indicates whether the Cloud Application has been redeployed from this machine
+        /// </summary>
+        public bool LocallyRedeployed { get; set; }
```
Comment (ganeshnj) [inline src/AWS.Deploy.Common/CloudApplication.cs:40]: This meant to be the current (logged in) user not "machine" in the comment. 

Also, it will make easier to understand if we rename it to `UpdatedByCurrentUser` IMO.

File: src/AWS.Deploy.Orchestration/LocalUserSettings/LocalUserSettingsEngine.cs [hunk 33]
```diff
@@ -0,0 +1,208 @@
...
+                var localUserSettings = await GetLocalUserSettings();
+                var localStacks = localUserSettings?.LastDeployedStacks?
+                    .FirstOrDefault(x => x.Exists(awsAccountId, awsRegion, projectName));
+
+                if (localUserSettings == null || localStacks == null)
+                    return;
+
+                var validStacks = deployedStacks.Intersect(localStacks.Stacks);
+
+                localStacks.Stacks = validStacks.ToList();
```
Comment (ganeshnj) [inline src/AWS.Deploy.Orchestration/LocalUserSettings/LocalUserSettingsEngine.cs:150]: Where is `validStacks` used? Possibly both statements can be removed.
```suggestion
```

**Review by ganeshnj — CHANGES_REQUESTED** [review]

File: src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestEngine.cs (hunk 30 above)
Comment (philasmar) [reply src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestEngine.cs:63]: They are not redundant. When we are deserializing them, if they don't exist in the file, they will be null. Ive ran into null exceptions multiple times which is why i added the check.

File: src/AWS.Deploy.Orchestration/LocalUserSettings/LocalUserSettingsEngine.cs (hunk 31 above)
Comment (philasmar) [reply src/AWS.Deploy.Orchestration/LocalUserSettings/LocalUserSettingsEngine.cs:57]: They are not redundant. When we are deserializing them, if they don't exist in the file, they will be null. Ive ran into null exceptions multiple times which is why i added the check.

File: src/AWS.Deploy.Common/CloudApplication.cs (hunk 32 above)
Comment (philasmar) [reply src/AWS.Deploy.Common/CloudApplication.cs]: Done

File: src/AWS.Deploy.Orchestration/LocalUserSettings/LocalUserSettingsEngine.cs (hunk 33 above)
Comment (philasmar) [reply src/AWS.Deploy.Orchestration/LocalUserSettings/LocalUserSettingsEngine.cs:150]: validStacks is being set to  localStacks.Stacks
So we are basically getting the intersection, setting it in the model and then writing the file back.

File: src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestEngine.cs (hunk 30 above)
Comment (ganeshnj) [reply src/AWS.Deploy.Common/DeploymentManifest/DeploymentManifestEngine.cs:63]: That's interesting, there must be some config that we are missing here. We can check it later.

File: src/AWS.Deploy.Orchestration/LocalUserSettings/LocalUserSettingsEngine.cs (hunk 33 above)
Comment (ganeshnj) [reply src/AWS.Deploy.Orchestration/LocalUserSettings/LocalUserSettingsEngine.cs:150]: Makes sense, it was hard to follow in one go.

**Review by ganeshnj — APPROVED** [review]

**Review by 96malhar — APPROVED** [review]
